### ISSUE TITLE
Security Issue related to non-changing defaults in the SUPERUSER_API_TOKEN

### ISSUE THREAD

**Issue by mosesrenegade** [netbox-community/netbox-docker#953]
### Current Behavior

During a security review of Netbox we have found that by default Netbox has several hardcoded defaults:

Username: admin
Password: admin

In our research, it appears almost all administrators that place their Netbox on the open internet change these defaults. We have however also noticed that there is a hardcoded SUPERUSER_API_TOKEN that is located here:

https://github.com/netbox-community/netbox-docker/blob/8b16b16e45ca88506521ca03de9320d07ec8e9ed/env/netbox.env#L34

But, we have also found that approximately 10-15% of all the publicly exposed Netbox systems on the internet do not have the API Token changed. We have found little to no documentation about the fact that this docker image ships with the default token in NetBox. We believe this is due to several reasons:

1. It is not listed in the documentation to do so
2. It is not recommended to do so
3. It is hardcoded and silent in the docker implementation.

The systems we have seen actually contain populated data and do not appear to be test systems, making this even more alarming.

### Expected Behavior

The recommendation that this value be "automatically" generated, even if simply using something as simple as: 

In the .env file:

SUPERUSER_API_TOKEN=$SUPERUSER_API_TOKEN

export SUPERUSER_API_RAND=$(LC_ALL=C tr -dc 'a-f0-9' </dev/urandom | head -c 25)
envsubst < .env

This would automatically generate a random token each time and insert it into the .env file. 

It would also be ideal to alert all administrators that have never used the API to check the default admin user API key to ensure that they change the API KEY to something other than the static value. We always recommend not placing a sensitive system like Netbox on the public internet, but this may be the design of the system itself.

### Docker Compose Version

Docker Compose version v2.13.0

### Docker Version

```text
Client:
 Cloud integration: v1.0.29
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 18:01:18 2022
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true
```

### The git Revision

39c7de4af41703108f8b3af7274bc65813c41799

### The git Status

```text
On branch release
Your branch is up to date with 'origin/release'.

nothing to commit, working tree clean
```

### Startup Command

docker-compose up -d

### NetBox Logs

```text
🧬 loaded config '/etc/netbox/config/configuration.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/extra.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/logging.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/plugins.py'
netbox-docker-netbox-1               | 💡 Superuser Username: admin, E-Mail: admin@example.com
netbox-docker-netbox-1               | ✅ Initialisation is done.
netbox-docker-netbox-1               | ⏳ Waiting for control socket to be created... (1/10)
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [warn] 7#7 Unit is running unprivileged, then it cannot use arbitrary user and group.
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 7#7 unit 1.27.0 started
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 20#20 discovery started
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [notice] 20#20 module: python 3.10.4 "/usr/lib/unit/modules/python3.10.unit.so"
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 7#7 controller started
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [notice] 7#7 process 20 exited with code 0
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 22#22 router started
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 22#22 OpenSSL 3.0.2 15 Mar 2022, 30000020
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 23#23 "netbox" prototype started
netbox-docker-netbox-1               | 2023/03/06 15:24:29 [info] 24#24 "netbox" application started
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/configuration.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/extra.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/logging.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/plugins.py'
netbox-docker-netbox-1               | ⚙️ Applying configuration from /etc/unit/nginx-unit.json
netbox-docker-netbox-1               | ✅ Unit configuration loaded successfully
netbox-docker-netbox-1               | 2023/03/06 15:24:31 [notice] 7#7 process 18 exited with code 0
netbox-docker-netbox-1               | 2023/03/06 15:24:37 [info] 37#37 "netbox" application started
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/configuration.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/extra.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/logging.py'
netbox-docker-netbox-1               | 🧬 loaded config '/etc/netbox/config/plugins.py'
netbox-docker-netbox-1               | 127.0.0.1 - - [06/Mar/2023:15:24:41 +0000] "GET /api/ HTTP/1.1" 200 469 "-" "curl/7.81.0"
netbox-docker-netbox-1               | 127.0.0.1 - - [06/Mar/2023:15:24:55 +0000] "GET /api/ HTTP/1.1" 200 469 "-" "curl/7.81.0"
```

### Content of docker-compose.override.yml

```yaml
version: '3.4'
services:
  netbox:
    ports:
    - 8000:8080
```

**Comment by mosesrenegade** [comment]
The following CVE has been reserved for tracking of this issue:

CVE-2023-27573

**Comment by kkthxbye-code** [comment]
> During a security review of Netbox we have found that by default **Netbox has several hardcoded defaults**:

As I see you have registered a CVE, please make sure that you make it very clear that this is not an issue with the netbox project, but the netbox-docker project. netbox-docker is a community driven project.

> We have found little to no documentation about the fact that this docker image ships with the default token in NetBox.

https://github.com/netbox-community/netbox-docker/wiki/Configuration#configure-for-production

https://github.com/netbox-community/netbox-docker/wiki/Deployment

> The recommendation that this value be "automatically" generated, even if simply using something as simple as:

While I agree that it is problematic, I think the solution should be to remove the placeholder value from the config and modify `docker-entrypoint.sh` to just skip creation of the token if not set. I see little value for the average user to create a superuser API token by default.

**Comment by mosesrenegade** [comment]
> As I see you have registered a CVE, please make sure that you make it very clear that this is not an issue with the netbox 
> project, but the netbox-docker project. netbox-docker is a community driven project.

I will let the MITRE team note this, that this is a community supported stack and not related to the Netbox core (or the company) product stack.

I do agree skipping the creation of an API token should probably be the default behavior, but it is not known if this can be done at all or if there was a reason that the key had to default set?

**Comment by cimnine** [comment]
> \[…\] if there was a reason that the key had to default set?

The main purpose of the NetBox Docker image for a long time was to assist development of services that rely on the NetBox API for their job. Hence, a static API token was necessary. When the popularity of NetBox Docker grew, plenty of folks started to use it for their production or production-like setups (e.g. home and community setups).

The wiki page on how to set up NetBox Docker for production(-like) use-cases [is among the oldest we have](https://github.com/netbox-community/netbox-docker/wiki/Configuration/64bc716107481adea28070d1b5ed8ea51917232b). Before, the same content was in our `README.md`. But that file got too large, so parts were moved to our wiki. Also, the wiki was way easier to adjust than opening a PR for the `README.md`.

> I think the solution should be to remove the placeholder value from the config and modify docker-entrypoint.sh to just skip creation of the token if not set.

I second this.

> \[…\] but it is not known if this can be done at all \[…\]

It can be done. We'll just mention the change in the release notes, no problem.

Which brings us to the elephant in the room: What is the sensible move for those instances already out there? One idea is to check in the entrypoint script whether the default API key is defined and print a warning. Or – more extreme – `exit 1`. (Perhaps with the exception that a certain (new) environment variable is defined). Now, I'm not sure how we'd go about that – programming wise.

Another major issue is that _someone_ would have to do it. Myself, I'm currently really busy with other tasks in my life. I'm not sure about the situation of our other contributors either. As said earlier:

> \[…\] netbox-docker is a community driven project.

**Comment by tobiasge** [comment]
Closed with #955
